crates/red_knot_python_semantic/src/types.rs

```diff
@@ -471,6 +471,154 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if this type and `other` have no common elements.
+    ///
+    /// Note: This function aims to have no false positives, but might return
+    /// wrong `false` answers in some cases.
+    pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        match (self, other) {
+            (Type::Never, _) | (_, Type::Never) => true,
+
+            (Type::Any, _) | (_, Type::Any) => false,
+            (Type::Unknown, _) | (_, Type::Unknown) => false,
+            (Type::Unbound, _) | (_, Type::Unbound) => false,
+            (Type::Todo, _) | (_, Type::Todo) => false,
+
+            (ty @ (Type::Function(..) | Type::Module(..) | Type::Class(..)), other)
+            | (other, ty @ (Type::Function(..) | Type::Module(..) | Type::Class(..))) => {
+                ty != other
+            }
+
+            (Type::Union(union), other) | (other, Type::Union(union)) => union
+                .elements(db)
+                .iter()
+                .all(|e| e.is_disjoint_from(db, other)),
+
+            (Type::Intersection(intersection), other)
+            | (other, Type::Intersection(intersection)) => {
+                if intersection
+                    .positive(db)
+                    .iter()
+                    .any(|p| p.is_disjoint_from(db, other))
+                {
+                    true
+                } else {
+                    // TODO we can do better here. For example:
+                    // X & ~Literal[1] is disjoint from Literal[1]
+                    false
+                }
+            }
+
+            (Type::Tuple(tuple), other) | (other, Type::Tuple(tuple)) => {
+                if let Type::Tuple(other_tuple) = other {
+                    if tuple.len(db) == other_tuple.len(db) {
+                        tuple
+                            .elements(db)
+                            .iter()
+                            .zip(other_tuple.elements(db).iter())
+                            .any(|(e1, e2)| e1.is_disjoint_from(db, *e2))
+                    } else {
+                        true
+                    }
+                } else {
+                    true
+                }
+            }
+
+            (
+                Type::None,
+                Type::IntLiteral(_)
+                | Type::BooleanLiteral(_)
+                | Type::StringLiteral(..)
+                | Type::LiteralString
+                | Type::BytesLiteral(..),
+            )
+            | (
+                Type::IntLiteral(_)
+                | Type::BooleanLiteral(_)
+                | Type::StringLiteral(..)
+                | Type::LiteralString
+                | Type::BytesLiteral(..),
```

Comment:
Just an oversight. I like to use `..` for matching on AST/Type enums because they tend to appear in *lots* of `match` statements. If I can be certain that changing the *structure* of a particular item will not require someone to revisit this pattern, I can make future refactorings easier by using `..` instead of `_`, `_, _`, ….